staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go

```diff
@@ -96,10 +97,29 @@ func NewTestable(
 	return newTestableController(informerFactory, flowcontrolClient, serverConcurrencyLimit, requestWaitLimit, obsPairGenerator, queueSetFactory)
 }
 
+// appendUnauthenticatedGroupIfNotInKnownGroups modifies the passed request
+// digest to append system:unauthenticated if the none of the user's groups
+// matches the list of known groups.
+func appendUnauthenticatedGroupIfNotInKnownGroups(requestDigest *RequestDigest) {
```

Comment:
curious why this was needed for beta but not alpha

Reply:
Along with graduation to beta, we also turn on APF by default in this PR -- the latter is the reason why we need this. Unfortunately, some integration tests constructed API server clients in a hacky way, which resulted in requests without any user groups. This isn't a problem in production because all requests are guaranteed to have either system:authenticated or system:unauthenticated.

Reply:
> Unfortunately, some integration tests constructed API server clients in a hacky way, which resulted in requests without any user groups.

Can we fix those tests to match reality, rather than changing production code to accomodate tests?

Reply:
We did try, but there are too many tests that do this in too many different ways. For what it's worth, assuming such tests to be system:unauthenticated is a similar baseline to no-APF -- such requests get matched to a catch-all flow schema.

Reply:
> We did try, but there are too many tests that do this in too many different ways.

Can we try harder? :)

I'd like to avoid iterating through all groups in production

Reply:
+1 for avoiding that in production

@adtac - how many such tests are?

Reply:
> Can we fix those tests to match reality, rather than changing production code to accomodate tests?

Strongly disagree--the fact that we hit these in tests means that nothing else in production enforces this invariant, which means we have no idea if production will violate this or not. And in fact we already found and fixed a problem with this while in alpha. There are probably more.

I don't know about iteration, but we probably already do that since flow schemas can match groups.

Reply:
I would like @deads2k to look at this though.

I'd personally prefer an approach that didn't modify the user request, I'm worried that it might not be a deep copy, the first version of this I saw last night didn't have that problem but I haven't looked at the current version carefully.

Reply:
>how many such tests are?

@wojtek-t virtually all integration tests do this. Looking at [this prow link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/96527/pull-kubernetes-integration/1327257366239907840), I estimate it at ~2000 tests, but the number of instantiations of the API server that are at fault should be in the mid hundreds, I think.

Reply:
>I'd like to avoid iterating through all groups in production

If performance is the concern, there is a way to do this without iterating through each group that we discussed last night -- instead of appending system:unauthenticated, we can select the catch-all flow schema by default. This approach would have virtually zero impact on performance. It is functionally the same, but less "elegant" and does not address the root cause.

Reply:
> @wojtek-t virtually all integration tests do this

That seems like a bug to fix, rather than work around here. Can you give details about the integration tests that are making requests that don't have an authenticated or unauthenticated group attached? The only issue I was aware of was around impersonation, and that was resolved in https://github.com/kubernetes/kubernetes/pull/94204

Reply:
> the number of instantiations of the API server that are at fault

in the cases I'm aware of, API server instantiations in integration tests use the same authentication filter as real API servers... if there's a path that results in a user without one of these groups, we need to know about it

Reply:
> we can select the catch-all flow schema if all matches fail.

Yeah, I would strongly prefer this.

> in the cases I'm aware of, API server instantiations in integration tests use the same authentication filter as real API servers... if there's a path that results in a user without one of these groups, we need to know about it

If our *tests* generate this condition and we didn't notice, then I tend to believe this "invariant" actually is in our heads only, and not the system. :) I don't think it should fall on the APF team to add in an invariant.

Reply:
Put it another way: fixing the tests to not do bogus things with the user would NOT give me confidence that we wouldn't break clusters in production. It's fine to fix the tests later, but we need this code to be defensive either way.

We already found one live problem in a real cluster where the "invariant" didn't hold, while in alpha! I now just plain don't believe this is really an invariant.